client_code/persistence.py

```diff
@@ -0,0 +1,173 @@
+
+
+class LinkedClass:
+    "A descriptor class for adding objects based on linked tables as attributes"
+
+    def __init__(self, cls, *args, linked_column=None, **kwargs):
+        self._linked_column = linked_column
+        self._cls = cls
+        self._args = args or []
+        self._kwargs = kwargs or {}
+
+    def __get__(self, instance, objtype=None):
+        if instance is None:
+            return self
+
+        return self._cls(
+            instance._store[self._linked_column], *self._args, **self._kwargs
+        )
+
+    def __set__(self, instance, value):
+        raise AttributeError(
+            "Linked Class instance is already set and cannot be changed"
+        )
+
+
+class PersistedClass:
+    key = None
+
+    @classmethod
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        cls._snake_name = _snakify(cls.__name__)
+        cls._cache = {}
+        for attr, value in cls.__dict__.items():
+            try:
+                is_persisted_class = issubclass(value, PersistedClass)
+            except TypeError:
+                is_persisted_class = False
+
+            if is_persisted_class:
+                setattr(cls, attr, LinkedClass(cls=value, linked_column=attr))
+
+    @classmethod
+    def search(cls, lazy=False, *args, **kwargs):
+        rows = anvil.server.call(f"search_{cls._snake_name}", *args, **kwargs)
+        if lazy:
+            return (cls(store=row) for row in rows)
+
+        result = [cls(store=row) for row in rows]
+        cls.cache.clear()
+        for obj in result:
+            cls._cache[getattr(obj, cls.key)] = obj
+        return result
+
+    @classmethod
+    def get(cls, key):
+        try:
+            return cls._cache[key]
+        except KeyError:
+            row = anvil.server.call(f"get_{cls._snake_name}", {cls.key: key})
+            obj = cls(store=row)
+            cls._cache[key] = obj
+            return obj
+
+    def __init__(self, store=None, *args, **kwargs):
+        self._store = store or {}
+        self._delta = {}
+
+    def __getattr__(self, key):
+        if self._delta and key in self._delta:
+            return self._delta[key]
+
+        try:
+            return self._store[key]
+        except (KeyError, anvil.tables.TableError):
+            raise AttributeError(key)
+
+    def __getitem__(self, key):
+        try:
+            return getattr(self, key)
+        except AttributeError:
+            raise KeyError(key)
+
+    def __setattr__(self, key, value):
+        is_private = key.startswith("_")
+        is_descriptor = hasattr(self.__class__, key) and hasattr(
+            getattr(self.__class__, key), "__set__"
+        )
+        if is_private or is_descriptor:
+            object.__setattr__(self, key, value)
+        else:
+            self._delta[key] = value
+
+    def __setitem__(self, key, value):
+        setattr(self, key, value)
+
+    def add(self, *args, **kwargs):
```

Comment:
I think it would be nice to document this. The api isn't obvious since it's an instance method so you need to create the object before adding it, and the server call should return the store.

Reply:
Looking at the docs, I think a bit of rewrite might not be a bad thing...